zerver/lib/upload.py

```diff
@@ -385,14 +386,32 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
     return uploaded_file_name, uploaded_file_size, content_type
 
 
+@contextlib.contextmanager
+def maybe_unset_proxy():
+    # https://github.com/boto/botocore/issues/2644 means that the IMDS
+    # request _always_ pulls from the environment.  If S3_SKIP_PROXY
+    # and S3_KEY are unset, or S3_SKIP_PROXY is True, then we need to
+    # temporarily remove the environment keys to let the IMDS request
+    # through.
+    old_environ = dict(os.environ)
+    if settings.S3_SKIP_PROXY == True or (settings.S3_SKIP_PROXY is None and not settings.S3_KEY):
+        os.environ["HTTP_proxy"] = ""
```

Comment:
This isn’t thread-safe. Maybe we can monkey-patch `botocore.utils.should_bypass_proxies = lambda url: True`?

Reply:
Pushed a monkeypatch version!